File: Material.hs
module Material where

import Ray
import Vec3
import Texture

import Prelude hiding ((<*>))

data Material = Lambertian Texture |
                Metal Texture Double | 
                DiffuseLight Texture deriving Show

scatter :: Material -> Ray -> Vec3 -> Vec3 -> Vec2 -> Maybe (Ray, Vec3)

scatter (Lambertian albedo) _ hitPosition hitNormal hitTexPosition =
    let p = hitPosition
        n = hitNormal

        -- Hack to create a somewhat random number
        (px, py, pz) = p
        (nx, ny, nz) = n
        kindaRand = 10000000.0 * (px + nx + py + ny + pz + nz)

        target = n <+> (randomInUnitSphere (round kindaRand))
        scattered = Ray { rayOrigin = p, rayDirection = normalize target }
    in Just (scattered, textureValue albedo hitPosition hitTexPosition) 

scatter (Metal albedo fuzz) rayIn hitPosition hitNormal hitTexPosition = 
    let rd = rayDirection rayIn
        p = hitPosition
        n = hitNormal
        
        -- Hack to create a somewhat random number
        (px, py, pz) = p
        (nx, ny, nz) = n
        kindaRand = 10000000.0 * (px + nx + py + ny + pz + nz)

        target = (reflect rd n) <+> ((randomInUnitSphere (round kindaRand)) <*> fuzz)
        scattered = Ray { rayOrigin = p, rayDirection = normalize target}
    in Just (scattered, textureValue albedo hitPosition hitTexPosition)

scatter (DiffuseLight _) _ _ _ _ = Nothing

materialEmitted :: Material -> Vec3 -> Vec3
materialEmitted (Lambertian _) _ = (0.0, 0.0, 0.0)
materialEmitted (Metal _ _) _ = (0.0, 0.0, 0.0)
materialEmitted (DiffuseLight t) p = textureValue t p (0.0, 0.0)

File: raytracer-cpp/src/material.cpp
#include "material.h"

ScatterResult LambertianMaterial::scatter(const Ray &ray, const Vec3 &position, const Vec3 &normal, const Vec2 &texture_coord) {
    ScatterResult result;

    Vec3 target = normal + Vec3::random_in_unit_sphere();
    result.ray = Ray(position, Vec3::normalize(target)); 
    result.color = albedo->value(texture_coord, position);
    result.did_scatter = true;

    return result;
}

Vec3 LambertianMaterial::emitted(const Vec3 &p) {
    return Vec3(0.0, 0.0, 0.0);
}

ScatterResult MetalMaterial::scatter(const Ray &ray, const Vec3 &position, const Vec3 &normal, const Vec2 &texture_coord) {
    ScatterResult result;

    Vec3 target = Vec3::reflect(ray.direction, normal);
    result.ray = Ray(position, Vec3::normalize(target)); 
    result.color = albedo->value(texture_coord, position);
    result.did_scatter = true;

    return result;
}

Vec3 MetalMaterial::emitted(const Vec3 &p) {
    return Vec3(0.0, 0.0, 0.0);
}

ScatterResult DiffuseLightMaterial::scatter(const Ray &ray, const Vec3 &position, const Vec3 &normal, const Vec2 &texture_coord) {
    ScatterResult result;
    result.did_scatter = false;
    return result;
}

Vec3 DiffuseLightMaterial::emitted(const Vec3 &p) {
    return albedo->value(Vec2(0.0, 0.0), p);
}
